File: python/examples/all_in_one_performance.py
import asyncio
import random
import time
from collections.abc import AsyncGenerator
from typing import Any

from camera_ui_rpc import ErrorCode, RPCClass, RPCException, ServiceConfig, create_rpc_client


@RPCClass
class PerformanceTimer:
    def __init__(self, name: str):
        self.name = name
        self.start_time = 0.0
        self.operations: list[tuple[str, float]] = []

    def start_operation(self, operation: str) -> "OperationTimer":
        return OperationTimer(self, operation)

    def record_operation(self, operation: str, duration: float) -> None:
        self.operations.append((operation, duration))
        print(f"  {operation}: {duration:.2f}ms")

    def print_summary(self) -> None:
        print(f"\n{self.name} Performance Summary:")
        print("=" * 50)
        total_time = sum(duration for _, duration in self.operations)
        for operation, duration in self.operations:
            percentage = (duration / total_time) * 100 if total_time > 0 else 0
            print(f"{operation:.<40} {duration:>8.2f}ms ({percentage:>5.1f}%)")
        print("=" * 50)
        print(f"{'Total Time:':.<40} {total_time:>8.2f}ms")
        print()


@RPCClass
class OperationTimer:
    def __init__(self, timer: PerformanceTimer, operation: str):
        self.timer = timer
        self.operation = operation
        self.start_time = 0.0

    def __enter__(self) -> "OperationTimer":
        self.start_time = time.perf_counter()
        return self

    def __exit__(self, *args: Any) -> None:
        duration = (time.perf_counter() - self.start_time) * 1000
        self.timer.record_operation(self.operation, duration)


small_data = "This is a small test message"
medium_data = bytes([i % 256 for i in range(5 * 1024 * 1024)])  # 5MB
large_data = bytes([i % 256 for i in range(10 * 1024 * 1024)])  # 10MB


@RPCClass
class TestService:
    async def echo(self, msg: str) -> str:
        return msg

    async def add(self, a: int, b: int) -> int:
        return a + b

    async def get_large_data(self) -> bytes:
        return large_data

    async def echo_data(self, data: bytes) -> bytes:
        return data

    async def generate_numbers(self, count: int) -> AsyncGenerator[int, None]:
        for i in range(count):
            yield i

    async def error_method(self, should_fail: bool) -> str:
        if should_fail:
            raise RPCException(ErrorCode.INTERNAL_ERROR, "Test error")
        return "Success"


async def test_all_features() -> None:
    print("Comprehensive RPC Performance Test\n")

    timer = PerformanceTimer("All-in-One Test")

    with timer.start_operation("Server Client Creation"):
        server = create_rpc_client(
            {
                "servers": ["nats://localhost:4222"],
                "name": "test-server",
                "auth": {"user": "server", "password": "server_password"},
            }
        )
        await server.connect()

    with timer.start_operation("Client Creation"):
        client = create_rpc_client(
            {
                "servers": ["nats://localhost:4222"],
                "name": "test-client",
                "auth": {"user": "server", "password": "server_password"},
            }
        )
        await client.connect()

    print("\n 1. Native Request/Reply")
    with timer.start_operation("Request Handler Setup & 100 calls"):

        async def echo_handler(data: Any) -> dict[str, Any]:
            return {"echo": data["msg"], "timestamp": int(time.perf_counter() * 1000)}

        unsub = await server.on_request("echo.*", echo_handler)
        await asyncio.sleep(0.05)  # Let subscription settle

        for i in range(100):
            result = await client.request("echo.test", {"msg": f"Message {i}"})
            assert result["echo"] == f"Message {i}"

        await unsub()

    print("\n 2. Register Handlers (RPC style)")
    with timer.start_operation("RPC Handler Setup & 100 calls"):

        async def handle_echo(msg: str) -> str:
            return f"Echo: {msg}"

        async def handle_add(a: int, b: int) -> int:
            return a + b

        handlers: dict[str, Any] = {
            "echo": handle_echo,
            "add": handle_add,
        }

        unsub_rpc = await server.register_handler("test", handlers)
        await asyncio.sleep(0.05)  # More time for handler setup

        test_proxy = client.create_proxy("test")

        for i in range(50):
            echo_result = await test_proxy.echo(f"Message {i}")
            assert echo_result == f"Echo: Message {i}"

            add_result = await test_proxy.add(i, i + 1)
            assert add_result == 2 * i + 1

        await unsub_rpc()

    print("\n 3. Large Data Transfer (Auto-Chunking)")
    with timer.start_operation("Large Data Transfer (10MB)"):

        async def handle_large() -> bytes:
            return large_data

        async def handle_echo_data(data: bytes) -> bytes:
            return data

        large_handlers: dict[str, Any] = {
            "get_large": handle_large,
            "echo_data": handle_echo_data,
        }

        unsub_large = await server.register_handler("data", large_handlers)
        await asyncio.sleep(0.05)

        data_proxy = client.create_proxy("data")

        result = await data_proxy.get_large()
        assert len(result) == len(large_data)

        echo_result = await data_proxy.echo_data(medium_data)
        assert echo_result == medium_data

        await unsub_large()

    print("\n 4. Channel Communication")
    with timer.start_operation("Channel Setup & 1000 messages"):
        server_channel = await server.channel("perf-channel")
        client_channel = await client.channel("perf-channel")

        messages_received: list[dict[str, Any]] = []

        def on_message(msg: dict[str, Any]) -> None:
            messages_received.append(msg)

        server_channel.on("message", on_message)
        await asyncio.sleep(0.05)

        for i in range(1000):
            await client_channel.send({"index": i, "data": f"Message {i}"})

        await asyncio.sleep(0.2)
        assert len(messages_received) == 1000

        await server_channel.close()
        await client_channel.close()

    print("\n 5. Private Channel Communication")
    with timer.start_operation("Private Channel Setup & 100 calls"):
        server_private = await server.private_channel("perf-private", "test-client")

        async def handle_private(data: dict[str, Any]) -> dict[str, Any]:
            return {"processed": data["value"] * 2}

        unsub_private = await server_private.on_request(handle_private)
        await asyncio.sleep(0.05)

        client_private = await client.private_channel("perf-private", "test-server")

        for i in range(100):
            result = await client_private.request({"value": i}, timeout=5000)
            assert result["processed"] == i * 2

        await unsub_private()
        await server_private.close()
        await client_private.close()

    print("\n 6. Service Creation & Discovery")
    with timer.start_operation("Service Setup"):
        service = await server.service.register_handler(
            ServiceConfig(name="perf-service", version="1.0.0", description="Performance test service"),
            TestService(),
        )

        await asyncio.sleep(0.1)  # Let service register

        monitor = client.service.monitor()
        services = await monitor.info("perf-service")
        assert len(services) > 0
        assert services[0].name == "perf-service"

    print("\n 7. Service Proxy Calls")
    with timer.start_operation("Service Proxy Creation & 50 calls"):
        calc_service = await client.create_service_proxy("perf-service", TestService)

        for i in range(50):
            result = await calc_service.add(i, i + 1)
            assert result == 2 * i + 1

    print("\n 8. Service Streaming")
    with timer.start_operation("Stream Processing (100 items)"):
        stream_sum = 0
        stream = calc_service.generate_numbers(100)

        async for value in stream:
            stream_sum += value

        assert stream_sum == sum(range(100))

    # Services don't support chunking, so large data goes via RPC
    print("\n 8b. Large Data via RPC Handler")
    with timer.start_operation("Get Large Data (10MB) via RPC"):

        async def handle_get_large_data() -> bytes:
            return large_data

        large_data_handlers: dict[str, Any] = {"getLargeData": handle_get_large_data}

        unsub_large_rpc = await server.register_handler("largedata", large_data_handlers)
        await asyncio.sleep(0.05)

        large_proxy = client.create_proxy("largedata")
        large_result = await large_proxy.getLargeData()
        assert len(large_result) == len(large_data)

        await unsub_large_rpc()

    print("\n 9. Concurrent Operations")
    with timer.start_operation("Concurrent Requests (500 parallel)"):

        async def concurrent_handler(data: dict[str, Any]) -> dict[str, Any]:
            return {"echo": data, "handled": True}

        unsub_concurrent = await server.on_request("echo.concurrent", concurrent_handler)
        await asyncio.sleep(0.05)

        async def concurrent_call(i: int) -> dict[str, Any]:
            return await client.request("echo.concurrent", {"index": i})

        tasks = [concurrent_call(i) for i in range(500)]
        results = await asyncio.gather(*tasks)
        assert len(results) == 500

        # Keep the handler for mixed workload test
        # await unsub_concurrent()

    print("\n 10. Error Handling")
    with timer.start_operation("Error Handling Test"):
        result = await calc_service.echo("test")
        assert result == "test"

        try:
            await calc_service.error_method(True)
            raise AssertionError("Should have raised exception")
        except RPCException as e:
            # The error code might come as a string "500" or the enum value
            assert str(e.code) == "500" or e.code == ErrorCode.INTERNAL_ERROR

    print("\n 11. Isolated Connection Proxy")
    with timer.start_operation("Isolated Proxy Test"):
        handlers_isolated: dict[str, Any] = {
            "echo": handle_echo,
            "add": handle_add,
        }
        unsub_isolated = await server.register_handler("test", handlers_isolated)
        await asyncio.sleep(0.05)

        isolated_proxy_with_close = client.create_proxy("test", isolated_connection=True)
        isolated_proxy = isolated_proxy_with_close.proxy

        for i in range(10):
            result = await isolated_proxy.echo(f"Isolated {i}")
            assert result == f"Echo: Isolated {i}"

        await isolated_proxy_with_close.close()
        await unsub_isolated()

    print("\n 12. Mixed Workload (Simulating Real Usage)")
    with timer.start_operation("Mixed Operations"):
        test_channel = await client.channel("mixed-channel")

        async def mixed_operation():
            ops: list[asyncio.Task[Any]] = []

            for _ in range(10):
                ops.append(asyncio.create_task(client.request("echo.concurrent", {"msg": "quick"})))

                ops.append(
                    asyncio.create_task(calc_service.add(random.randint(1, 100), random.randint(1, 100)))
                )

                ops.append(asyncio.create_task(test_channel.send({"event": "mixed", "data": "test"})))

            # Service calls don't support auto-chunking, so use echo to avoid "maximum payload exceeded"
            ops.append(asyncio.create_task(calc_service.echo("test data")))

            await asyncio.gather(*ops)

        rounds: list[asyncio.Task[Any]] = []
        for _ in range(5):
            rounds.append(asyncio.create_task(mixed_operation()))

        await asyncio.gather(*rounds)
        await test_channel.close()

    print("\nCleanup")
    with timer.start_operation("Cleanup"):
        await unsub_concurrent()
        await service.stop()
        await client.disconnect()
        await server.disconnect()

    timer.print_summary()


async def main() -> None:
    try:
        await test_all_features()
        print("All tests completed successfully!")
    except Exception as e:
        print(f"Test failed: {e}")
        import traceback

        traceback.print_exc()
        raise


if __name__ == "__main__":
    import uvloop

    uvloop.run(main())

File: node/examples/all-in-one-performance.ts
import { RPCClass } from '../src/decorators.js';
import { createRPCClient, ERROR_CODES, RPCException } from '../src/index.js';

class PerformanceTimer {
  private name: string;
  private operations: { operation: string; duration: number }[] = [];

  constructor(name: string) {
    this.name = name;
  }

  startOperation(operation: string): OperationTimer {
    return new OperationTimer(this, operation);
  }

  recordOperation(operation: string, duration: number): void {
    this.operations.push({ operation, duration });
    console.log(`  ${operation}: ${duration.toFixed(2)}ms`);
  }

  printSummary(): void {
    console.log(`\n${this.name} Performance Summary:`);
    console.log('='.repeat(50));
    const totalTime = this.operations.reduce((sum, op) => sum + op.duration, 0);

    for (const { operation, duration } of this.operations) {
      const percentage = totalTime > 0 ? (duration / totalTime) * 100 : 0;
      const padding = '.'.repeat(Math.max(40 - operation.length, 0));
      console.log(`${operation}${padding} ${duration.toFixed(2).padStart(8)}ms (${percentage.toFixed(1).padStart(5)}%)`);
    }

    console.log('='.repeat(50));
    const totalPadding = '.'.repeat(Math.max(40 - 'Total Time:'.length, 0));
    console.log(`Total Time:${totalPadding} ${totalTime.toFixed(2).padStart(8)}ms`);
    console.log();
  }
}

class OperationTimer {
  private timer: PerformanceTimer;
  private operation: string;
  private startTime = 0;

  constructor(timer: PerformanceTimer, operation: string) {
    this.timer = timer;
    this.operation = operation;
  }

  start(): void {
    this.startTime = performance.now();
  }

  end(): void {
    const duration = performance.now() - this.startTime;
    this.timer.recordOperation(this.operation, duration);
  }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const smallData = 'This is a small test message';
const mediumData = Buffer.alloc(5 * 1024 * 1024); // 5MB
const largeData = Buffer.alloc(10 * 1024 * 1024); // 10MB

for (let i = 0; i < mediumData.length; i++) {
  mediumData[i] = i % 256;
}
for (let i = 0; i < largeData.length; i++) {
  largeData[i] = i % 256;
}

@RPCClass
class TestService {
  async echo(msg: string): Promise<string> {
    return msg;
  }

  async add(a: number, b: number): Promise<number> {
    return a + b;
  }

  async getLargeData(): Promise<Buffer> {
    return largeData;
  }

  async echoData(data: Buffer): Promise<Buffer> {
    return data;
  }

  async *generateNumbers(count: number): AsyncGenerator<number> {
    for (let i = 0; i < count; i++) {
      yield i;
    }
  }

  async errorMethod(shouldFail: boolean): Promise<string> {
    if (shouldFail) {
      throw new RPCException(ERROR_CODES.INTERNAL_ERROR, 'Test error');
    }
    return 'Success';
  }
}

async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function testAllFeatures(): Promise<void> {
  console.log('Comprehensive RPC Performance Test\n');

  const timer = new PerformanceTimer('All-in-One Test');

  const serverOp = timer.startOperation('Server Client Creation');
  serverOp.start();
  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    name: 'test-server',
    auth: { user: 'server', password: 'server_password' },
  });
  await server.connect();
  serverOp.end();

  const clientOp = timer.startOperation('Client Creation');
  clientOp.start();
  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    name: 'test-client',
    auth: { user: 'server', password: 'server_password' },
  });
  await client.connect();
  clientOp.end();

  console.log('\n1. Native Request/Reply');
  const test1Op = timer.startOperation('Request Handler Setup & 100 calls');
  test1Op.start();

  const unsub1 = await server.onRequest('echo.*', async (data: any) => {
    return { echo: data.msg, timestamp: Date.now() };
  });
  await sleep(50); // Let subscription settle

  for (let i = 0; i < 100; i++) {
    const result = await client.request('echo.test', { msg: `Message ${i}` });
    if (result.echo !== `Message ${i}`) {
      throw new Error('Echo mismatch');
    }
  }

  unsub1();
  test1Op.end();

  console.log('\n2. Register Handlers (RPC style)');
  const test2Op = timer.startOperation('RPC Handler Setup & 100 calls');
  test2Op.start();

  const handlers = {
    echo: async (msg: string) => `Echo: ${msg}`,
    add: async (a: number, b: number) => a + b,
  };

  const unsub2 = await server.registerHandler('test', handlers);
  await sleep(50);

  const testProxy = client.createProxy<typeof handlers>('test');

  for (let i = 0; i < 50; i++) {
    const echoResult = await testProxy.echo(`Message ${i}`);
    if (echoResult !== `Echo: Message ${i}`) {
      throw new Error('Echo result mismatch');
    }

    const addResult = await testProxy.add(i, i + 1);
    if (addResult !== 2 * i + 1) {
      throw new Error('Add result mismatch');
    }
  }

  await unsub2();
  test2Op.end();

  console.log('\n3. Large Data Transfer (Auto-Chunking)');
  const test3Op = timer.startOperation('Large Data Transfer (10MB)');
  test3Op.start();

  const largeHandlers = {
    getLarge: async () => largeData,
    echoData: async (data: Buffer) => data,
  };

  const unsub3 = await server.registerHandler('data', largeHandlers);
  await sleep(50);

  const dataProxy = client.createProxy<typeof largeHandlers>('data');

  const result = await dataProxy.getLarge();
  if (result.length !== largeData.length) {
    throw new Error('Large data size mismatch');
  }

  const echoResult = await dataProxy.echoData(mediumData);
  const echoBuffer = Buffer.isBuffer(echoResult) ? echoResult : Buffer.from(echoResult);
  if (!echoBuffer.equals(mediumData)) {
    throw new Error('Echo data mismatch');
  }

  await unsub3();
  test3Op.end();

  console.log('\n4. Channel Communication');
  const test4Op = timer.startOperation('Channel Setup & 1000 messages');
  test4Op.start();

  const serverChannel = await server.channel('perf-channel');
  const clientChannel = await client.channel('perf-channel');

  const messagesReceived: any[] = [];

  serverChannel.on('message', (msg) => {
    messagesReceived.push(msg);
  });
  await sleep(50);

  for (let i = 0; i < 1000; i++) {
    await clientChannel.send({ index: i, data: `Message ${i}` });
  }

  await sleep(200);
  if (messagesReceived.length !== 1000) {
    throw new Error(`Expected 1000 messages, got ${messagesReceived.length}`);
  }

  await serverChannel.close();
  await clientChannel.close();
  test4Op.end();

  console.log('\n5. Private Channel Communication');
  const test5Op = timer.startOperation('Private Channel Setup & 100 calls');
  test5Op.start();

  const serverPrivate = await server.privateChannel('perf-private', 'test-client');

  const unsub5 = await serverPrivate.onRequest(async (data: any) => {
    return { processed: data.value * 2 };
  });
  await sleep(50);

  const clientPrivate = await client.privateChannel('perf-private', 'test-server');

  for (let i = 0; i < 100; i++) {
    const result = await clientPrivate.request({ value: i }, 5000);
    if (result.processed !== i * 2) {
      throw new Error('Private channel result mismatch');
    }
  }

  unsub5();
  await serverPrivate.close();
  await clientPrivate.close();
  test5Op.end();

  console.log('\n6. Service Creation & Discovery');
  const test6Op = timer.startOperation('Service Setup');
  test6Op.start();

  const service = await server.service.registerHandler(
    {
      name: 'perf-service',
      version: '1.0.0',
      description: 'Performance test service',
    },
    new TestService(),
  );

  await sleep(100);

  const services = server.service.getAllInfo();
  if (services.length === 0) {
    throw new Error('Service not found');
  }
  if (services[0].name !== 'perf-service') {
    throw new Error('Service name mismatch');
  }

  test6Op.end();

  console.log('\n7. Service Proxy Calls');
  const test7Op = timer.startOperation('Service Proxy Creation & 50 calls');
  test7Op.start();

  const calcService = await client.createServiceProxy<TestService>('perf-service');

  for (let i = 0; i < 50; i++) {
    const result = await calcService.add(i, i + 1);
    if (result !== 2 * i + 1) {
      throw new Error('Service add result mismatch');
    }
  }

  test7Op.end();

  console.log('\n8. Service Streaming');
  const test8Op = timer.startOperation('Stream Processing (100 items)');
  test8Op.start();

  let streamSum = 0;
  for await (const value of calcService.generateNumbers(100)) {
    streamSum += value;
  }

  const expectedSum = (99 * 100) / 2; // sum of 0 to 99
  if (streamSum !== expectedSum) {
    throw new Error(`Stream sum mismatch: ${streamSum} !== ${expectedSum}`);
  }

  test8Op.end();

  // Services don't support chunking; use an RPC handler for large data
  console.log('\n8b. Large Data via RPC Handler');
  const test8bOp = timer.startOperation('Get Large Data (10MB) via RPC');
  test8bOp.start();

  const largeDataHandlers = {
    getLargeData: async () => largeData,
  };

  const unsub8b = await server.registerHandler('largedata', largeDataHandlers);
  await sleep(50);

  const largeProxy = client.createProxy<typeof largeDataHandlers>('largedata');
  const largeResult = await largeProxy.getLargeData();
  if (largeResult.length !== largeData.length) {
    throw new Error('Large data RPC size mismatch');
  }

  await unsub8b();
  test8bOp.end();

  console.log('\n9. Concurrent Operations');
  const test9Op = timer.startOperation('Concurrent Requests (500 parallel)');
  test9Op.start();

  const unsub9 = await server.onRequest('echo.concurrent', async (data: any) => {
    return { echo: data, handled: true };
  });
  await sleep(50);

  const concurrentCall = async (i: number) => {
    return await client.request('echo.concurrent', { index: i });
  };

  const tasks = Array.from({ length: 500 }, (_, i) => concurrentCall(i));
  const results = await Promise.all(tasks);
  if (results.length !== 500) {
    throw new Error('Concurrent results count mismatch');
  }

  // Keep the handler for mixed workload test
  test9Op.end();

  console.log('\n10. Error Handling');
  const test10Op = timer.startOperation('Error Handling Test');
  test10Op.start();

  const successResult = await calcService.echo('test');
  if (successResult !== 'test') {
    throw new Error('Echo success mismatch');
  }

  try {
    await calcService.errorMethod(true);
    throw new Error('Should have raised exception');
  } catch (e: any) {
    // Services return HTTP-like error codes (500), not ERROR_CODES
    if (!(e instanceof RPCException) || (e.code !== '500' && e.code !== ERROR_CODES.INTERNAL_ERROR)) {
      throw new Error(`Unexpected error: ${e}`);
    }
  }

  test10Op.end();

  console.log('\n11. Isolated Connection Proxy');
  const test11Op = timer.startOperation('Isolated Proxy Test');
  test11Op.start();

  const handlersIsolated = {
    echo: async (msg: string) => `Echo: ${msg}`,
    add: async (a: number, b: number) => a + b,
  };

  const unsub11 = await server.registerHandler('test', handlersIsolated);
  await sleep(50);

  const isolatedProxyWithClose = client.createProxy<typeof handlersIsolated>('test', { isolatedConnection: true });

  for (let i = 0; i < 10; i++) {
    const result = await isolatedProxyWithClose.proxy.echo(`Isolated ${i}`);
    if (result !== `Echo: Isolated ${i}`) {
      throw new Error('Isolated echo mismatch');
    }
  }

  await isolatedProxyWithClose.close();
  await unsub11();
  test11Op.end();

  console.log('\n12. Mixed Workload (Simulating Real Usage)');
  const test12Op = timer.startOperation('Mixed Operations');
  test12Op.start();

  const testChannel = await client.channel('mixed-channel');

  const mixedOperation = async () => {
    const ops: Promise<any>[] = [];

    for (let i = 0; i < 10; i++) {
      ops.push(client.request('echo.concurrent', { msg: 'quick' }));

      ops.push(calcService.add(Math.floor(Math.random() * 100), Math.floor(Math.random() * 100)));

      ops.push(testChannel.send({ event: 'mixed', data: 'test' }));
    }

    // Service calls don't support auto-chunking; use echo to avoid "maximum payload exceeded"
    ops.push(calcService.echo('test data'));

    await Promise.all(ops);
  };

  const rounds: Promise<void>[] = [];
  for (let i = 0; i < 5; i++) {
    rounds.push(mixedOperation());
  }

  await Promise.all(rounds);
  await testChannel.close();
  test12Op.end();

  console.log('\nCleanup');
  const cleanupOp = timer.startOperation('Cleanup');
  cleanupOp.start();

  unsub9(); // Clean up the concurrent handler
  await service.stop();
  await client.disconnect();
  await server.disconnect();

  cleanupOp.end();

  timer.printSummary();
}

async function main(): Promise<void> {
  try {
    await testAllFeatures();
    console.log('All tests completed successfully!');
  } catch (error) {
    console.error('Test failed:', error);
    process.exit(1);
  }
}

main().catch(console.error);
